.github/workflows/combined_format_lint_test_mypy.yml

```
@@ -28,6 +27,104 @@ jobs:
+            tr " " "\n" |
+            grep -E "\.txt$|\.json$|\.csv$"  |
+            tr "\n" " ")" >> "$GITHUB_ENV"
+      
+      - name: Check for Unauthorized Changes
+        run: |
+          paths_to_protect=(
+            'RUFAS/scenario_manager.py'
+          )
+          unauthorized_changes=false
+          
+          for path in "${paths_to_protect[@]}"; do
+            if (echo ${{ env.FILES }}; echo ${{ env.OTHER_FILES }}) | grep -q "$path"; then
+              unauthorized_changes=true
+              unauthorized_paths+="$path, "
+            fi
+          done
+  
+          if $unauthorized_changes; then
+            echo "Unauthorized changes detected in protected files: ${unauthorized_paths%, }"
+            echo "Review and remove these changes if they are not intended."
+            exit 1
+          else
+            echo "No unauthorized changes detected."
+          fi
+      
+      - name: Format code with Black
+        if: ${{ env.FILES != '' && github.ref != 'refs/heads/main' }}
+        run: |
+          black ${{ env.FILES }}
+
+      - name: Commit and push formatting changes
+        if: ${{ env.FILES != '' && github.ref != 'refs/heads/main' }}
+        env:
+          BRANCH_NAME: ${{ github.head_ref || github.ref }}
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          git config --global user.name 'github-actions[bot]'
+          git config --global user.email 'github-actions[bot]@users.noreply.github.com'
+          git add -A
+          git commit -m "Apply Black Formatting" --allow-empty
+          git remote set-url origin https://${GITHUB_TOKEN}@github.com/${{ github.repository }}.git
+          CLEAN_BRANCH_NAME=${BRANCH_NAME/refs\/heads\//}
+          git push -f origin HEAD:$CLEAN_BRANCH_NAME
+      
+  
+  lint_test_type_check_unauthorized_changes:
+    needs: format_code
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+        with:
+          persist-credentials: false
+          fetch-depth: 0
+      
+      - name: Fetch latest commit SHA
+        id: fetch-latest-sha
+        run: |
+          BRANCH_NAME="${{ github.head_ref || github.ref_name }}"
+          echo "Branch name: $BRANCH_NAME"
+          LATEST_SHA=$(git rev-parse origin/"$BRANCH_NAME")
+          echo "Latest SHA: $LATEST_SHA"
+          echo "LATEST_SHA=$LATEST_SHA" >> $GITHUB_ENV
+          echo "::set-output name=sha::$LATEST_SHA"
+      
+      - name: Checkout the latest commit
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ steps.fetch-latest-sha.outputs.sha }}
+      
+      - name: Set up Python 3.10
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.10"
+
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install coverage mypy
+          if [ -f requirements.txt ]; then pip install -r requirements.txt; fi 
```

```suggestion
          if [ -f requirements.txt ]; then pip install -r requirements.txt; fi 
```
Isn't the installation of Mypy and Coverage handled by this install?

---

dependencies don't persist through github jobs even within the same action. But you're right, we can do away with the independent mypy and coverage install since they're in the requirements.txt file. Good catch!